### PR TITLE
SPSDK fix and Pynitrokey 0.4.31 -> 0.4.34

### PR DESCRIPTION
**Pull request by frogamic**
###### Description of changes

Relax more dependencies in python3Packages.spsdk as it is broken:
- pycryptodome:  expected <v3.17, nixos provides v3.17 ([changelog](https://github.com/Legrandin/pycryptodome/blob/master/Changelog.rst))
- typing-extensions: expected <v4.5, nixos provides v4.5 ([changelog](https://github.com/python/typing_extensions/blob/main/CHANGELOG.md))

Update pynitrokey to v0.4.34 ([changelog](https://github.com/Nitrokey/pynitrokey/releases))
- Re-order dependencies to match pyproject.toml in pynitrokey
- Relax dependencies in pynitrokey:
  - typing-extensions: expected v4.3.x, nixos provides v4.5 ([changelog](https://github.com/python/typing_extensions/blob/main/CHANGELOG.md))
  - python-dateutil: expected v2.7.x, nixos provides 2.8.2 ([changelog](https://github.com/dateutil/dateutil/releases))

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

closes https://github.com/NixOS/nixpkgs/pull/222465

### CODE REVIEW

File: pkgs/tools/security/pynitrokey/default.nix [hunk 1]
```diff
@@ -1,47 +1,67 @@
...
 buildPythonApplication rec {
   pname = "pynitrokey";
-  version = "0.4.31";
+  version = "0.4.34";
   format = "flit";
 
   src = fetchPypi {
     inherit pname version;
-    sha256 = "sha256-nqw5wUzQxKCBzYBRhqB6v7WWrF1Ojf8z6Kf1YUA9+wU=";
+    sha256 = "sha256-lMXoDkNiAmGb6e4u/vZMcmXUclwW402YUGihLjWIr+U=";
```
Comment (dotlambda) [inline pkgs/tools/security/pynitrokey/default.nix:32]:
```suggestion
    hash = "sha256-lMXoDkNiAmGb6e4u/vZMcmXUclwW402YUGihLjWIr+U=";
```

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -39239,7 +39239,7 @@ with pkgs;
 
   xrq = callPackage ../applications/misc/xrq { };
 
-  pynitrokey = callPackage ../tools/security/pynitrokey { };
+  pynitrokey = python3Packages.callPackage ../tools/security/pynitrokey { };
```
Comment (dotlambda) [inline pkgs/top-level/all-packages.nix:39242]:
```suggestion
  pynitrokey = callPackage ../tools/security/pynitrokey { };
```
We want to make it possible to override dependencies.

File: pkgs/tools/security/pynitrokey/default.nix [hunk 3]
```diff
@@ -1,47 +1,67 @@
...
     urllib3
-    cffi
-    cbor
-    nkdfu
-    fido2
     tlv8
-  ];
-
-  nativeBuildInputs = [
-    pythonRelaxDepsHook
```
Comment (dotlambda) [inline pkgs/tools/security/pynitrokey/default.nix:37]: No need to move this.

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (frogamic) [reply pkgs/top-level/all-packages.nix]: I didn't realise using callPackage over python3Packages.callPackage was preferential.

**Comment by panicgh** [comment]
LGTM and works (I tried fw update and OPT/secrets app) with my Nitrokey 3 (with latest beta firmware).

**Comment by panicgh** [comment]
Since #223242 has been merged, you may consider to rebase you branch and apply https://github.com/panicgh/nixpkgs/commit/6ab58ace6c2381b9d842b68543abd064574049ee. It adds libnitrokey, which makes nitropy support e.g. Nitrokey Pro devices (I've tested this with real HW).

**Comment by frogamic** [comment]
I tested @panicgh changes (without hardware) and pynitrokey does find libnitrokey and attempt to use it

**Comment by panicgh** [comment]
The commit message "python310Packages.spsdk: fix build by relaxing versions" could be misleading, because it not only affects python310Packages. Previous commits have used "python3Packages.spsdk" prefix (but also "python310Packages.spsdk" occurs at least once).

**Review by panicgh — APPROVED** [review]
LGTM. I successfully used `nitropy` with a Nitrokey Pro and a Nitrokey 3 (with FW v1.3.0-alpha.20230320).

<details>
<summary>Nitrokey Pro</summary>

```console
$ ./results/pynitrokey/bin/nitropy pro list
Command line tool to interact with Nitrokey devices 0.4.34
:: 'Nitrokey Pro' keys:
Using env variable supplied path: /nix/store/yw3rjrq7wiladh6crzx8nafsyljykgrg-libnitrokey-3.8/lib
Nitrokey Pro-XXXX 1-9:1.0
```
</details>

<details>
<summary>Nitrokey 3</summary>

```console
$ ./results/pynitrokey/bin/nitropy nk3 test
Command line tool to interact with Nitrokey devices 0.4.34
Found 1 Nitrokey 3 device(s):
- Nitrokey 3 at /dev/hidraw0

Running tests for Nitrokey 3 at /dev/hidraw0

[1/4]	uuid     	UUID query              	SUCCESS  	XXX
[2/4]	version  	Firmware version query  	SUCCESS  	v1.3.0-alpha.20230320
[3/4]	status   	Device status           	SUCCESS  	Status(init_status=<InitStatus.0: 0>, ifs_blocks=42, efs_blocks=478)
Please press the touch button on the device ...
Please press the touch button on the device ...
[4/4]	fido2    	FIDO2                   	SUCCESS  	

4 tests, 4 successful, 0 skipped, 0 failed

Summary: 1 device(s) tested, 1 successful, 0 failed
```
</details>

 Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>pynitrokey</li>
    <li>pynitrokey.dist</li>
    <li>python310Packages.spsdk</li>
    <li>python310Packages.spsdk.dist</li>
    <li>python311Packages.spsdk</li>
    <li>python311Packages.spsdk.dist</li>
  </ul>
</details>

**Pull request merged**